api/rootcerts.go

```diff
@@ -0,0 +1,157 @@
+
+package api
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+)
+
+// CertConfig determines where LoadCACerts will load certificates from.
+type certConfig struct {
+	// CAFile is a path to a PEM-encoded certificate file or bundle.
+	CAFile string
+
+	// CACertificate is a PEM-encoded certificate or bundle.
+	CACertificate []byte
+
+	// CAPath is a path to a directory populated with PEM-encoded certificates.
+	CAPath string
+}
+
+// ConfigureTLS sets up the RootCAs on the provided tls.Config based on the CertConfig specified.
+func configureTLS(t *tls.Config, c *certConfig) error {
+	if t == nil {
+		return nil
+	}
+
+	// Greedily load all possible CA sources into the pool.
+	pool, err := loadCACerts(c)
+	if err != nil {
+		return err
+	}
+	t.RootCAs = pool
+	return nil
+}
+
+// LoadCACerts loads a CertPool from all available CA sources in CertConfig.
+func loadCACerts(c *certConfig) (*x509.CertPool, error) {
+	// Start with system CAs.
+	pool, err := loadSystemCAs()
+	if err != nil {
+		pool = x509.NewCertPool()
+	}
+
+	// Load from CAFile, if specified.
+	if c.CAFile != "" {
+		if err := appendCAFile(pool, c.CAFile); err != nil {
+			return nil, err
+		}
+	}
+
+	// Load from CACertificate, if specified.
+	if len(c.CACertificate) != 0 {
+		if err := appendCertificate(pool, c.CACertificate); err != nil {
+			return nil, err
+		}
+	}
+
+	// Load from CAPath, if specified.
+	if c.CAPath != "" {
+		if err := appendCAPath(pool, c.CAPath); err != nil {
+			return nil, err
+		}
+	}
+
+	return pool, nil
+}
+
+// AppendCAFile loads a single PEM-encoded file and appends it to the provided CertPool.
+func appendCAFile(pool *x509.CertPool, caFile string) error {
+	pem, err := os.ReadFile(caFile)
+	if err != nil {
+		return fmt.Errorf("Error loading CA File: %w", err)
+	}
+
+	ok := pool.AppendCertsFromPEM(pem)
+	if !ok {
+		return fmt.Errorf("Error loading CA File: Couldn't parse PEM in: %s", caFile)
+	}
+
+	return nil
+}
+
+// AppendCertificate appends an in-memory PEM-encoded certificate or bundle to the provided CertPool.
+func appendCertificate(pool *x509.CertPool, ca []byte) error {
+	ok := pool.AppendCertsFromPEM(ca)
+	if !ok {
+		return errors.New("Error appending CA: Couldn't parse PEM")
+	}
+	return nil
+}
+
+func appendCAPath(pool *x509.CertPool, caPath string) error {
+	walkFn := func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if info.IsDir() {
+			return nil
+		}
+
+		pemData, err := os.ReadFile(path)
+		if err != nil {
+			return fmt.Errorf("Error loading file from CAPath: %w", err)
+		}
+
+		// Decode the PEM blocks and only append certificate blocks
+		var block *pem.Block
+		rest := pemData
+		for {
+			block, rest = pem.Decode(rest)
+			if block == nil {
+				break
+			}
+
+			// Ensure the PEM block is a certificate
+			if block.Type != "CERTIFICATE" {
+				fmt.Printf("Ignoring non-certificate PEM block found in %s\n", path)
```

Comment:
Hmmm.... How I dislike PEM some days :-) 

The use of `fmt.Printf` I'd remove: (semgrep fails on it -- I think you can run it with `make docker-semgrep-ci`) https://github.com/openbao/openbao/actions/runs/11141741694/job/30963245478?pr=574 -- philosophically, `api/` is a library which may be called in various contexts, so I wouldn't always expect this to be seen, but I agree it isn't an error; I'd just ignore it.

That said...

The `block.Type != "CERTIFICATE"` is also worth changing. There's [the `ParsePEMBundle`](https://github.com/openbao/openbao/blob/main/sdk/helper/certutil/helpers.go#L252-L309) function in `sdk/helper/certutil` that handles this, IMO, correctly with opportunistic parsing:

https://github.com/openbao/openbao/blob/e6ed492e570d7783c02ae7556e3f8fdba9f96ac5/sdk/helper/certutil/helpers.go#L285-L289

(In particular, there's a bunch of legacy header types you can sometimes run into... Like `X509 CERTIFICATE` and others).

Reply:
Having the `Printf` statement in production code didn't feel right, and I appreciate the test that checks for it and the suggestion to remove it. I’ve also updated the certificate parsing to use `x509.ParseCertificates` as suggested. I'm not too familiar with all the helper functions yet, but I'll take some time to look through them and get more acquainted.